### PR TITLE
[libc++] Fix failure with GCC 15 (#117319)

### LINKED ISSUE #117319: [libc++] libc++ doesn't compile with gcc 15
libc++ doesn't compile with GCC 15 because it comes with new GCC ```__decay``` built-in which is incompatible with libc++.
[similar issue](https://github.com/llvm/llvm-project/issues/91831)
[the bug file](https://github.com/llvm/llvm-project/blob/main/libcxx/include/__type_traits/decay.h#L28)

### PR DESCRIPTION
**Pull request by killcerr**
Fixes #117319

### CODE REVIEW

**Comment by frederick-vs-ja** [comment]
Hi! I think it would be better to associate this PR withe the issue to be fixed using [GitHub's close/fix/resolve syntax](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

**Comment by killcerr** [comment]
> Hi! I think it would be better to associate this PR withe the issue to be fixed using [GitHub's close/fix/resolve syntax](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).



I already change it.Thanks your comment.

**Review by philnik777 — CHANGES_REQUESTED** [review]
We don't support trunk GCC and this is definitely not the only problem. I don't think it makes sense to add an opt-out in a single place when we know that things are still broken.

**Comment by killcerr** [comment]
> We don't support trunk GCC and this is definitely not the only problem. I don't think it makes sense to add an opt-out in a single place when we know that things are still broken.

Thanks, should i close this pull request?

File: libcxx/include/__type_traits/decay.h [hunk 1]
```diff
@@ -25,7 +25,7 @@
 
 _LIBCPP_BEGIN_NAMESPACE_STD
 
-#if __has_builtin(__decay)
+#if __has_builtin(__decay) && !defined(_LIBCPP_COMPILER_GCC)
```
Comment (ldionne) [inline libcxx/include/__type_traits/decay.h:28]: How is the GCC-provided builtin incompatible with Clang's?

**Review by ldionne — COMMENTED** [review]
@killcerr Is that the only issue you encountered?

File: libcxx/include/__type_traits/decay.h (hunk 1 above)
Comment (philnik777) [reply libcxx/include/__type_traits/decay.h:28]: GCC doesn't mangle the builtin.

**Comment by killcerr** [comment]
> @killcerr Is that the only issue you encountered?

Yes, I met it when I complie GCC 15.

File: libcxx/include/__type_traits/decay.h (hunk 1 above)
Comment (ldionne) [reply libcxx/include/__type_traits/decay.h:28]: Ah, so do we need to do something like

```
template <class _Tp>
struct __decay_impl {
  using type = __builtin_decay_t(_Tp);
};

template <class _Tp>
using __decay_t = typename __decay_impl<_Tp>::type;
```

?

Comment (killcerr) [reply libcxx/include/__type_traits/decay.h:28]: [similar issue](https://github.com/llvm/llvm-project/issues/91831)
I use the same way to fix this issue.

Comment (philnik777) [reply libcxx/include/__type_traits/decay.h:28]: Yes, something like that. I'd probably just use `decay` as the base (i.e. `using __decay_t = decay<T>::type`) for GCC.

**Pull request closed without merging**